server/session/handler_enchanting.go

```diff
@@ -0,0 +1,341 @@
+			// the enchanting table.
+			continue
+		}
+
+		// Also make sure that the enchantment is compatible with the item.
+		if book || enchant.CompatibleWithItem(it) {
+			// Now iterate through each possible level of the enchantment.
+			for i := enchant.MaxLevel(); i > 0; i-- {
+				// Use the level to calculate the minimum and maximum costs for this enchantment.
+				if cost >= enchant.MinCost(i) && cost <= enchant.MaxCost(i) {
+					// If the cost is within the bounds, add the enchantment to the list of available enchantments.
+					availableEnchants = append(availableEnchants, item.NewEnchantment(enchant, i))
+					break
+				}
+			}
+		}
+	}
+	if len(availableEnchants) == 0 {
+		// No available enchantments, so we can't really do much here.
+		return nil
+	}
+
+	// Now we need to select the enchantments.
+	selectedEnchants := make([]item.Enchantment, 0, len(availableEnchants))
+
+	// Select the first enchantment using a weighted random algorithm, favouring enchantments that have a higher weight.
+	// These weights are based on the enchantment's rarity, with common and uncommon enchantments having a higher weight
+	// than rare and very rare enchantments.
+	enchant := weightedRandomEnchantment(random, availableEnchants)
+	selectedEnchants = append(selectedEnchants, enchant)
+
+	// Remove the selected enchantment from the list of available enchantments, so we don't select it again.
+	ind := sliceutil.Index(availableEnchants, enchant)
+	availableEnchants = slices.Delete(availableEnchants, ind, ind+1)
+
+	// Based on the cost, select a random amount of additional enchantments.
+	for random.Intn(50) <= cost {
+		// Ensure that we don't have any conflicting enchantments. If so, remove them from the list of available
+		// enchantments.
+		lastEnchant := selectedEnchants[len(selectedEnchants)-1]
+		if availableEnchants = sliceutil.Filter(availableEnchants, func(enchant item.Enchantment) bool {
+			return lastEnchant.Type().CompatibleWithEnchantment(enchant.Type())
+		}); len(availableEnchants) == 0 {
+			// We've exhausted all available enchantments.
+			break
+		}
+
+		// Select another enchantment using the same weighted random algorithm.
+		enchant = weightedRandomEnchantment(random, availableEnchants)
+		selectedEnchants = append(selectedEnchants, enchant)
+
+		// Remove the selected enchantment from the list of available enchantments, so we don't select it again.
+		ind = sliceutil.Index(availableEnchants, enchant)
+		availableEnchants = slices.Delete(availableEnchants, ind, ind+1)
+
+		// Halve the cost, so we have a lower chance of selecting another enchantment.
+		cost /= 2
+	}
+	return selectedEnchants
+}
+
+// searchBookshelves searches for nearby bookshelves around the position passed, and returns the amount found.
+func searchBookshelves(w *world.World, pos cube.Pos) (shelves int) {
+	for x := -1; x <= 1; x++ {
+		for z := -1; z <= 1; z++ {
+			for y := 0; y <= 1; y++ {
+				if z != 0 || x != 0 {
```

Comment:
Would try to nest this as little as possible. I'd invert this check `z == 0 && x == 0` and `continue` so we can move the code up a level.

Reply:
Inverted.